client/client.go

```diff
@@ -803,53 +780,147 @@ func (c *Client) Call(ctx context.Context, procedure string, options wamp.Dict,
+// will be at all) will be returned using progcb callback and the final result will be
+// returned using finalrescb callback
+//
+// This function returns a callback function that should be called with payload data chunks
+// To finish a progressive call the last callback MUST be called with isLast = true
+func (c *Client) CallProgressive(ctx context.Context, procedure string, options wamp.Dict, args wamp.List, kwargs wamp.Dict, finalrescb ResultHandler, progcb ProgressHandler) (SendProgressiveData, error) { //nolint:lll
+	if !c.Connected() {
+		return nil, ErrNotConn
+	}
+
+	if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeatureProgressiveCalls) {
+		return nil, ErrProgCallNotSupportedByRouter
+	}
+
+	if options == nil {
+		options = wamp.Dict{}
+	}
+
+	options[wamp.OptProgress] = true
+
+	receiveProgress := false
+
+	// If caller is willing to receive progressive results, create a channel to
+	// receive these on.  Then, start a goroutine to receive progressive
+	// results and call the callback for each.
+	var progChan chan *wamp.Result
+	var progDone chan struct{}
+	if progcb != nil {
+		receiveProgress = true
+		options[wamp.OptReceiveProgress] = receiveProgress
+		progChan = make(chan *wamp.Result)
+		progDone = make(chan struct{})
+		go func() {
+			for result := range progChan {
+				progcb(result)
 			}
+			close(progDone)
+		}()
+	}
+
+	if finalrescb == nil {
+		return nil, errors.New("no result handler provided")
+	}
+
+	id := c.idGen.Next()
+	c.expectReply(id)
+	message := &wamp.Call{
+		Request:   id,
+		Procedure: wamp.URI(procedure),
+		Options:   options,
+	}
+
+	err := c.prepareCallPayloadMessage(message, options, args, kwargs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.sess.Send(message)
+
+	// As we can not wait for final result and must return control to the client code
+	// we wait for result in goroutine
+	go func() {
+		// Wait to receive RESULT message.
+		msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)
+
+		// Finish handling any remaining progressive results before returning the
+		// final result.
+		if progcb != nil {
+			close(progChan)
+			<-progDone
+		}
+
+		if err != nil {
+			finalrescb(nil, err)
+			return
+		}
+
+		switch msg := msg.(type) {
+		case *wamp.Result:
+
+			err := c.prepareCallResultMessage(msg)
 
 			if err != nil {
-				return nil, err
+				finalrescb(nil, err)
+				return
 			}
 
-			msg.Arguments = args
-			msg.ArgumentsKw = kwargs
+			finalrescb(msg, nil)
+		case *wamp.Error:
+			finalrescb(nil, RPCError{msg, procedure})
+		default:
+			finalrescb(nil, unexpectedMsgError(msg, wamp.RESULT))
 		}
-		return msg, nil
-	case *wamp.Error:
-		return nil, RPCError{msg, procedure}
-	default:
-		return nil, unexpectedMsgError(msg, wamp.RESULT)
+	}()
+
+	// now we need to return callback for feeding the progressive data
+	sendProgressiveData := func(options wamp.Dict, args wamp.List, kwargs wamp.Dict, isLast bool) error {
```

Comment:
A slightly different approach to the pull model we discussed in the call. which also addresses the problem with the client code knowing when is the final call. Let's say we have the API like so:

```
type SendProgressHandler func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error)

// CallProgressive makes progressive calls to the procedure corresponding to
// the given URI.
//
func (c *Client) CallProgressive(ctx context.Context, procedure string, prog SendProgressHandler, progcb ProgressHandler) (*wamp.Result, error) { 
	...
}
```

When the caller returns `options[wamp.OptProgress] = true` in the return arguments from `SendProgressHandler`, this means that it's a intermediate invocation. The lack of `"progress"=true` option means it's the final invocation. So in this scenario, the progressive call like this:

```
progCb := func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error) {
	return wamp.List{"foo"}, nil, nil, nil
}
res, err := c.CallProgressive(ctx, "com.foo.bar", callProg, nil)
```

is equivalent to a simple call:

```
res, err := c.Call(ctx, "com.foo.bar", wamp.List{"foo"}, nil, nil, nil)
```

while this would be a progressive call with 2 invocations, where 2nd invocation is final (and still carries data):

```
cnt := 0
progCb := func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error) {
	if cnt == 0 { 
		cnt++
		return wamp.List{"the end"}, nil, wamp.Dict{wamp.OptProgress: true}, nil
	}
	return wamp.List{"the end"}, nil, nil, nil
}
res, err := c.CallProgressive(ctx, "com.foo.bar", callProg, nil)
```